### PR TITLE
Succinct graph representations

### PR DESCRIPTION
**Pull request by vigna**
This PR adds to jgrapht-unimi-dsi two graph representations based on succinct data structures.

The WebGraph adapters already provide ways to use succinct representation (e.g., EFGraph), but the implementations in this PR are modeled after the sparse representations of JGraphT—nodes and arcs are represented by integers and numbered starting from zero (they should be usable from Python).

Unfortunately, JGraphT's architecture clashes a bit with the succinct representation. For example, getEdgeSource() and getEdgeTarget() have to make twice the same expensive call. The result is that while the graphs are about 5 times smaller, access is 5 times slower, too.

UPDATE: My claims about speed are quite wrong.

The problem is that I (stupidly) quickly tested enumeration time for the whole arc set, which is not a good idea because in these graphs the arc set is trivial.

More accurate testing shows that that the directed version is 50% slower than SparseIntDirectedGraph when enumerating successors, and it is 2-3 times faster when checking adjacency (but see #1042 — the speed test in this case might be meaningless). This figures will vary with the density of the graph.

The undirected implementation is unfortunately significantly slower (5 times slower when enumerating successors). Adjacency, however, is about 150 times faster. Is there some reason why adjacency in SparseIntUndirectedGraph is so slow?

### CODE REVIEW

**Comment by d-michail** [comment]
By adjacency I guess that you mean `getAllEdges(source, target)`.  The problem is that we do not keep the list of neighbors sorted, in order to perform binary search. This is one of my TODO items.

**Comment by d-michail** [comment]
I will try to do this after #1029 gets merged.

**Comment by vigna** [comment]
> By adjacency I guess that you mean `getAllEdges(source, target)`. The problem is that we do not keep the list of neighbors sorted, in order to perform binary search. This is one of my TODO items.

Yes, or even just containsEdge(source, target). Maybe it's a linear search now? That would explain the timings.

I got a few interesting ideas and I'm going to partially rewrite this PR. My goal is to have size 4-5 times smaller than the sparse implementation and faster accessors. Let's see whether I can get there :).

**Comment by d-michail** [comment]
Yes, it is currently linear in the size of the neighborhood. `containsEdge` is similar or maybe just a wrapper for `getAllEdges()`, which means that both are linear.

**Comment by vigna** [comment]
OK, this is probably the most sensible approach. There is an implementation mimicking the sparse one which is quite slow, and one using pairs as edges that is an order of magnitude faster. I'm still doing some speed tests and I have to review the docs but it looks pretty usable. Footprint is 3 to 10 times less than the sparse implementation, depending on density.

**Comment by vigna** [comment]
OK, I think this is ready to merge if you like it. I have written extensive Javadocs as the tradeoffs between the two different kind of implementations might not be trivial to understand.

I don't know how that might be difficult, but having a bridge to Python for implementations using IntIntPair for edges might be very interesting—the succinct implementations using that instead of Integer are an order of magnitude faster.

**Comment by vigna** [comment]
BTW, do you guys think there's some value in a constructor for directed graphs that encodes only outgoing arcs? The space would be halved, but of course you woudn't get incoming arcs, similarly to the forward-only constructor of the WebGraph adapters.

**Comment by d-michail** [comment]
Yes, I did this recently for the `SparseIntDirectGraph`.

**Comment by vigna** [comment]
Ok, I'll do it ASAP.

**Comment by vigna** [comment]
I exploited also your new constructors using suppliers of edges of streams.

BTW, is there any reason why sparse representations are not serializable? I've been reliably storing and loading such instances without problems (just adding `implements Serializable`) to estimate their footprint.

Since it takes some time to build them I think it would be a useful feature. And people could easily publish graphs using that format.

**Comment by vigna** [comment]
Is there anything more I should do?

Once this is released, I was thinking about making part of the LAW graph database (say, graphs with <2B nodes) in this format. It would make it possible to test easily JGraphT on large graphs even with relatively little memory.

**Comment by d-michail** [comment]
Looks good. I will wait a bit to see if John has any comments.

File: jgrapht-unimi-dsi/src/main/java/org/jgrapht/sux4j/AbstractSuccinctDirectedGraph.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+ * First, we store the monotone lists of cumulative outdegrees and indegrees.
+ *
+ * <p>
+ * Then, we store the outgoing edges <var>x</var>&nbsp;&rarr;&nbsp;<var>y</var> as a monotone
+ * sequence using the encoding <var>x</var>2<sup>&lceil;log&nbsp;<var>n</var>&rceil;</sup> +
+ * <var>y</var>. At that point the <var>k</var>-th edge can be obtained by retrieving the
+ * <var>k</var>-th element of the sequence and some bit shifting (the encoding
+ * <var>x</var><var>n</var> + <var>y</var> would be slightly more compact, but much slower to
+ * decode). Since we know the list of cumulative outdegrees, we know which range of indices
+ * corresponds to the edges outgoing from each vertex. If we need to now whether
```
Comment (jsichi) [inline jgrapht-unimi-dsi/src/main/java/org/jgrapht/sux4j/AbstractSuccinctDirectedGraph.java:51]: typo: "know"

File: jgrapht-unimi-dsi/src/main/java/org/jgrapht/sux4j/SuccinctIntUndirectedGraph.java [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+ * Elias&ndash;Fano representation of monotone sequences} to represent the positions of ones in the
+ * (linearized) adjacency matrix of the graph. Instances are serializable and thread safe.
+ *
+ * <p>
+ * If the vertex set is compact (i.e., vertices are numbered from 0 consecutively), space usage will
+ * be close to the information-theoretical lower bound (typically, a few times smaller than a
+ * {@link SparseIntUndirectedGraph}).
+ *
+ * <p>
+ * {@linkplain org.jgrapht.GraphIterables#outgoingEdgesOf(Object) Enumeration of edges} is is very
```
Comment (jsichi) [inline jgrapht-unimi-dsi/src/main/java/org/jgrapht/sux4j/SuccinctIntUndirectedGraph.java:60]: typo "is is"

File: jgrapht-unimi-dsi/src/main/java/org/jgrapht/sux4j/SuccinctDirectedGraph.java [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+ * be close to twice the information-theoretical lower bound (typically, a few times smaller than a
+ * {@link SparseIntDirectedGraph}). If you {@link #SuccinctDirectedGraph(Graph, boolean) drop
+ * support for incoming edges} the space will close to the information-theoretical lower bound .
+ *
+ * <p>
+ * All accessors are very fast. {@linkplain org.jgrapht.Graph#containsEdge(Object) Adjacency tests}
+ * are very fast and happen in almost constant time.
+ *
+ * <p>
+ * {@link SuccinctDirectedGraph} is a much slower implementation with a similar footprint using
```
Comment (jsichi) [inline jgrapht-unimi-dsi/src/main/java/org/jgrapht/sux4j/SuccinctDirectedGraph.java:65]: I think this is supposed to reference SuccintIntDirectedGraph instead?

**Review by jsichi — APPROVED** [review]
Just a few doc nits.  Now we can proudly say, "JGraphT Sux!"

**Comment by vigna** [comment]
Thank you for the check! All fixed. 

BTW, once I talked with a guy who wouldn't use the library because of the word pun 😂.

**Comment by d-michail** [comment]
Just merged, thanks!

**Comment by vigna** [comment]
Great! Any planned release date for 1.5.1?

**Comment by jkinable** [comment]
> Great! Any planned release date for 1.5.1?

yes, yesterday :)

**Comment by vigna** [comment]
I don't know if this is the intended behavior, and I don't know if this might be an Ivy problem, but adding the artifact `jgrapht` to my `ivy.xml` dependencies does not bring in any jar. I have to manually point Ivy to, say, `jgraph-unimi-dsi`.

**Comment by d-michail** [comment]
Yes, you need to explicitly request the module that you want to use. The core library is the jgrapht-core.

**Comment by jkinable** [comment]
@vigna @d-michail 
Similar to the webgraphs in #1002 , can we include a demo/tutorial on the usage/purpose of succinct graphs? Key points:

- When to use a succinct graph representation: advantages/disadvantages and some examples of practical applications
- What is the advantage of a succinct graph representation over existing graph implementations, in particular, how does a succinct graph representation compare against (1) the 'standard' jgrapht graphs, (2) the specialized implementations in the `jgrapht-opt` module, (3) the webgraph representation. It should be clear for the user which implementation he/she should pick.

Prior to this PR, I personally hadn't heard about 'succinct' graphs. This might be used in a very specific field? We should try to make this accessible to 'standard' users. Also, from the wiki I understand that succinct graphs are incredibly space efficient. The only (?) reason to make graph storage this efficient is when you intend to store a massive graph (massive in terms of nr of edges/vertices). The same question applies here: do the JGraphT algorithms perform well enough on those massive graphs? Here we can obviously limit ourselves to algorithms you would reasonably expect to execute on those graphs, e.g. you would not compute an exact TSP on a graph with 20MM vertices.

**Comment by vigna** [comment]
I tried to explain all this in the Javadoc, but we can rephrase this somewhere else. The point is that we already have succinct representation in WebGraph, but these are more targeted at JGraphT and in particular to the Python bridge. We plan to distribute all our datasets with less than 2^31 arcs in serialized succinct JGraphT form directly, so users can just load and use them (less friction than with adapters).

Succinct data structures are fairly recent, with progress in implementations starting in the mid-2000. For example, Facebook's graph and text index are all stored using partitioned Elias-Fano, a succinct data structure (you can see some public code [here](https://github.com/facebook/folly/blob/master/folly/experimental/EliasFanoCoding.h), but what they actually use is more complex). Lucene has a [succinct Elias-Fano encoder](https://lucene.apache.org/core/4_8_0/core/org/apache/lucene/util/packed/EliasFanoEncoder.html).

The reason to use succinct graphs is simply that you can analyze in core memory much larger graphs. Access is asymptotically the same of a redundant format, with constant factors that can be large or small depending on the implementation.

**Comment by vigna** [comment]
We're having a look at the wiki with Dimitrios. Where would you think would be sensible to put information about this (and WebGraph adapters)?

**Comment by jsichi** [comment]
The correct place would be in the user guide (edit docs/guide-templates/UserOverview.md):

https://jgrapht.org/guide/UserOverview#graph-adapters

A new page (with code examples) linked from here would be best.

**Pull request merged**